File: DishForge.cs
// DishForge.cs - Генератор случайных блюд на C# (CLI + WinForms)
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Windows.Forms;

namespace DishForge
{
    public class Dish
    {
        public string Name { get; set; }
        public string Type { get; set; }
        public string Cuisine { get; set; }
        public int Time { get; set; }
        public string Difficulty { get; set; }
        public string Description { get; set; }
        public List<string> Ingredients { get; set; }
        public List<string> Instructions { get; set; }
    }

    public class Forge
    {
        public List<Dish> Dishes { get; set; } = new List<Dish>();
        public List<string> Favorites { get; set; } = new List<string>();
        private const string DataFile = "dishes.json";
        private const string FavFile = "favorites.json";

        public void Load()
        {
            if (File.Exists(DataFile))
            {
                try
                {
                    string json = File.ReadAllText(DataFile);
                    var dishes = JsonSerializer.Deserialize<List<Dish>>(json);
                    if (dishes != null) Dishes = dishes;
                }
                catch { }
            }
            if (!Dishes.Any()) { Dishes = DefaultDishes(); Save(); }
            if (File.Exists(FavFile))
            {
                try
                {
                    string json = File.ReadAllText(FavFile);
                    var favs = JsonSerializer.Deserialize<List<string>>(json);
                    if (favs != null) Favorites = favs;
                }
                catch { }
            }
        }

        public void Save()
        {
            File.WriteAllText(DataFile, JsonSerializer.Serialize(Dishes, new JsonSerializerOptions { WriteIndented = true }));
            File.WriteAllText(FavFile, JsonSerializer.Serialize(Favorites, new JsonSerializerOptions { WriteIndented = true }));
        }

        private List<Dish> DefaultDishes()
        {
            return new List<Dish>
            {
                new Dish { Name = "Овсянка с ягодами", Type = "breakfast", Cuisine = "russian", Time = 10, Difficulty = "easy",
                    Description = "Полезный завтрак из овсяных хлопьев с ягодами и мёдом.",
                    Ingredients = new List<string>{"Овсяные хлопья 100г", "Молоко 200мл", "Ягоды 50г", "Мёд 1 ст.л."},
                    Instructions = new List<string>{"Вскипятить молоко", "Добавить хлопья, варить 5 мин", "Добавить ягоды и мёд"} },
                new Dish { Name = "Паста Карбонара", Type = "lunch", Cuisine = "italian", Time = 25, Difficulty = "medium",
                    Description = "Классическая итальянская паста с беконом и пармезаном.",
                    Ingredients = new List<string>{"Паста 200г", "Бекон 100г", "Яйца 2 шт", "Пармезан 50г", "Сливки 100мл"},
                    Instructions = new List<string>{"Отварить пасту", "Обжарить бекон", "Смешать яйца, сыр и сливки", "Соединить с пастой"} }
            };
        }

        public Dish Generate(string type, string cuisine, int? maxTime, string difficulty)
        {
            var filtered = Dishes.AsEnumerable();
            if (!string.IsNullOrEmpty(type)) filtered = filtered.Where(d => d.Type == type);
            if (!string.IsNullOrEmpty(cuisine)) filtered = filtered.Where(d => d.Cuisine == cuisine);
            if (maxTime.HasValue) filtered = filtered.Where(d => d.Time <= maxTime.Value);
            if (!string.IsNullOrEmpty(difficulty)) filtered = filtered.Where(d => d.Difficulty == difficulty);
            var list = filtered.ToList();
            if (!list.Any()) return null;
            return list[new Random().Next(list.Count)];
        }

        public List<Dish> GenerateMenu(int count, string type, string cuisine, int? maxTime, string difficulty)
        {
            var available = Dishes.AsEnumerable();
            if (!string.IsNullOrEmpty(type)) available = available.Where(d => d.Type == type);
            if (!string.IsNullOrEmpty(cuisine)) available = available.Where(d => d.Cuisine == cuisine);
            if (maxTime.HasValue) available = available.Where(d => d.Time <= maxTime.Value);
            if (!string.IsNullOrEmpty(difficulty)) available = available.Where(d => d.Difficulty == difficulty);
            var list = available.ToList();
            if (list.Count < count) count = list.Count;
            var rnd = new Random();
            return list.OrderBy(x => rnd.Next()).Take(count).ToList();
        }

        public bool AddFavorite(string name)
        {
            if (Favorites.Contains(name)) return false;
            Favorites.Add(name);
            Save();
            return true;
        }

        public bool RemoveFavorite(string name)
        {
            if (Favorites.Remove(name)) { Save(); return true; }
            return false;
        }

        public List<Dish> GetFavorites()
        {
            return Dishes.Where(d => Favorites.Contains(d.Name)).ToList();
        }

        public string FormatDish(Dish d)
        {
            var lines = new List<string>();
            lines.Add($"🍽️ {d.Name}");
            lines.Add($"  Тип: {d.Type}");
            lines.Add($"  Кухня: {d.Cuisine}");
            lines.Add($"  Время: {d.Time} мин");
            lines.Add($"  Сложность: {d.Difficulty}");
            lines.Add($"  Описание: {d.Description}");
            lines.Add("  Ингредиенты:");
            d.Ingredients.ForEach(ing => lines.Add($"    - {ing}"));
            lines.Add("  Приготовление:");
            for (int i = 0; i < d.Instructions.Count; i++) lines.Add($"    {i+1}. {d.Instructions[i]}");
            return string.Join("\n", lines);
        }
    }

    class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            if (args.Length > 0 && args[0] == "--gui")
            {
                Application.EnableVisualStyles();
                Application.Run(new DishForgeGUI());
                return;
            }
            var forge = new Forge();
            forge.Load();
            if (args.Length == 0) { InteractiveMode(forge); return; }
            try
            {
                string cmd = args[0];
                switch (cmd)
                {
                    case "generate":
                        string type = null, cuisine = null, difficulty = null;
                        int? maxTime = null;
                        for (int i = 1; i < args.Length; i++)
                        {
                            if (args[i] == "--type") type = args[++i];
                            else if (args[i] == "--cuisine") cuisine = args[++i];
                            else if (args[i] == "--time") maxTime = int.Parse(args[++i]);
                            else if (args[i] == "--difficulty") difficulty = args[++i];
                        }
                        var dish = forge.Generate(type, cuisine, maxTime, difficulty);
                        if (dish != null) Console.WriteLine(forge.FormatDish(dish));
                        else Console.WriteLine("❌ Блюда не найдены");
                        break;
                    case "menu":
                        int count = 3; type = null; cuisine = null; difficulty = null; maxTime = null;
                        for (int i = 1; i < args.Length; i++)
                        {
                            if (args[i] == "--count") count = int.Parse(args[++i]);
                            else if (args[i] == "--type") type = args[++i];
                            else if (args[i] == "--cuisine") cuisine = args[++i];
                            else if (args[i] == "--time") maxTime = int.Parse(args[++i]);
                            else if (args[i] == "--difficulty") difficulty = args[++i];
                        }
                        var menu = forge.GenerateMenu(count, type, cuisine, maxTime, difficulty);
                        if (menu.Any())
                        {
                            Console.WriteLine($"🍽️ Меню из {menu.Count} блюд:");
                            for (int i = 0; i < menu.Count; i++)
                            {
                                var d = menu[i];
                                Console.WriteLine($"\n{i+1}. {d.Name} ({d.Type}, {d.Cuisine}, {d.Time} мин, {d.Difficulty})");
                                Console.WriteLine($"   {d.Description}");
                            }
                        }
                        else Console.WriteLine("❌ Недостаточно блюд");
                        break;
                    case "list":
                        type = null; cuisine = null;
                        for (int i = 1; i < args.Length; i++)
                        {
                            if (args[i] == "--type") type = args[++i];
                            else if (args[i] == "--cuisine") cuisine = args[++i];
                        }
                        var dishes = forge.Dishes;
                        if (!string.IsNullOrEmpty(type)) dishes = dishes.Where(d => d.Type == type).ToList();
                        if (!string.IsNullOrEmpty(cuisine)) dishes = dishes.Where(d => d.Cuisine == cuisine).ToList();
                        dishes.ForEach(d => Console.WriteLine($"{d.Name} ({d.Type}, {d.Cuisine}, {d.Time} мин)"));
                        break;
                    case "favorites":
                        if (args.Length > 2 && args[2] == "add" && args.Length > 3)
                        {
                            string name = args[3];
                            if (forge.AddFavorite(name)) Console.WriteLine($"✅ '{name}' добавлен в избранное");
                            else Console.WriteLine($"⚠️ '{name}' уже в избранном или не существует");
                        }
                        else if (args.Length > 2 && args[2] == "remove" && args.Length > 3)
                        {
                            string name = args[3];
                            if (forge.RemoveFavorite(name)) Console.WriteLine($"✅ '{name}' удалён из избранного");
                            else Console.WriteLine($"❌ '{name}' не найден в избранном");
                        }
                        else
                        {
                            var favs = forge.GetFavorites();
                            if (favs.Any())
                            {
                                Console.WriteLine("⭐ Избранное:");
                                favs.ForEach(d => Console.WriteLine($"  {d.Name} ({d.Type}, {d.Cuisine})"));
                            }
                            else Console.WriteLine("Избранное пусто");
                        }
                        break;
                    default: InteractiveMode(forge); break;
                }
            }
            catch (Exception e) { Console.WriteLine($"Ошибка: {e.Message}"); }
        }

        static void InteractiveMode(Forge forge)
        {
            while (true)
            {
                Console.WriteLine("\n🍳 DishForge - Генератор блюд (интерактивный)");
                Console.WriteLine("1. Сгенерировать случайное блюдо");
                Console.WriteLine("2. Сгенерировать меню");
                Console.WriteLine("3. Список всех блюд");
                Console.WriteLine("4. Избранное");
                Console.WriteLine("0. Выход");
                Console.Write("Выберите действие: ");
                string choice = Console.ReadLine();
                switch (choice)
                {
                    case "0": return;
                    case "1":
                        Console.Write("Тип (Enter пропустить): ");
                        string t = Console.ReadLine(); if (string.IsNullOrEmpty(t)) t = null;
                        Console.Write("Кухня (Enter пропустить): ");
                        string c = Console.ReadLine(); if (string.IsNullOrEmpty(c)) c = null;
                        Console.Write("Макс. время (мин, Enter пропустить): ");
                        string timeStr = Console.ReadLine();
                        int? mt = string.IsNullOrEmpty(timeStr) ? (int?)null : int.Parse(timeStr);
                        Console.Write("Сложность (easy/medium/hard, Enter пропустить): ");
                        string diff = Console.ReadLine(); if (string.IsNullOrEmpty(diff)) diff = null;
                        var dish = forge.Generate(t, c, mt, diff);
                        if (dish != null) Console.WriteLine(forge.FormatDish(dish));
                        else Console.WriteLine("❌ Блюда не найдены");
                        break;
                    case "2":
                        Console.Write("Количество блюд (по умолчанию 3): ");
                        string cntStr = Console.ReadLine();
                        int count = string.IsNullOrEmpty(cntStr) ? 3 : int.Parse(cntStr);
                        Console.Write("Тип (Enter пропустить): ");
                        string type = Console.ReadLine(); if (string.IsNullOrEmpty(type)) type = null;
                        Console.Write("Кухня (Enter пропустить): ");
                        string cuisine = Console.ReadLine(); if (string.IsNullOrEmpty(cuisine)) cuisine = null;
                        Console.Write("Макс. время (мин, Enter пропустить): ");
                        string timeStr2 = Console.ReadLine();
                        int? maxTime = string.IsNullOrEmpty(timeStr2) ? (int?)null : int.Parse(timeStr2);
                        Console.Write("Сложность (Enter пропустить): ");
                        string difficulty = Console.ReadLine(); if (string.IsNullOrEmpty(difficulty)) difficulty = null;
                        var menu = forge.GenerateMenu(count, type, cuisine, maxTime, difficulty);
                        if (menu.Any())
                        {
                            Console.WriteLine($"🍽️ Меню из {menu.Count} блюд:");
                            for (int i = 0; i < menu.Count; i++)
                            {
                                var d = menu[i];
                                Console.WriteLine($"\n{i+1}. {d.Name} ({d.Type}, {d.Cuisine}, {d.Time} мин, {d.Difficulty})");
                                Console.WriteLine($"   {d.Description}");
                            }
                        }
                        else Console.WriteLine("❌ Недостаточно блюд");
                        break;
                    case "3":
                        Console.Write("Тип (Enter пропустить): ");
                        string filterType = Console.ReadLine();
                        Console.Write("Кухня (Enter пропустить): ");
                        string filterCuisine = Console.ReadLine();
                        var dishes = forge.Dishes;
                        if (!string.IsNullOrEmpty(filterType)) dishes = dishes.Where(d => d.Type == filterType).ToList();
                        if (!string.IsNullOrEmpty(filterCuisine)) dishes = dishes.Where(d => d.Cuisine == filterCuisine).ToList();
                        dishes.ForEach(d => Console.WriteLine($"  {d.Name} ({d.Type}, {d.Cuisine}, {d.Time} мин)"));
                        break;
                    case "4":
                        var favs = forge.GetFavorites();
                        Console.WriteLine("⭐ Избранное:");
                        if (favs.Any()) favs.ForEach(d => Console.WriteLine($"  {d.Name} ({d.Type}, {d.Cuisine})"));
                        else Console.WriteLine("  Пусто");
                        Console.WriteLine("\n1. Добавить в избранное");
                        Console.WriteLine("2. Удалить из избранного");
                        Console.WriteLine("3. Назад");
                        Console.Write("Выберите: ");
                        string sub = Console.ReadLine();
                        if (sub == "1")
                        {
                            Console.Write("Название блюда: ");
                            string name = Console.ReadLine();
                            if (forge.AddFavorite(name)) Console.WriteLine($"✅ '{name}' добавлен");
                            else Console.WriteLine($"⚠️ '{name}' уже в избранном или не существует");
                        }
                        else if (sub == "2")
                        {
                            Console.Write("Название блюда: ");
                            string name = Console.ReadLine();
                            if (forge.RemoveFavorite(name)) Console.WriteLine($"✅ '{name}' удалён");
                            else Console.WriteLine($"❌ '{name}' не найден");
                        }
                        break;
                    default: Console.WriteLine("Неверный выбор");
                }
            }
        }
    }

    // ========== GUI ==========
    public class DishForgeGUI : Form
    {
        private Forge forge = new Forge();
        private ComboBox typeBox, cuisineBox, diffBox;
        private TextBox timeBox;
        private TextBox resultBox;
        private Dish currentDish;

        public DishForgeGUI()
        {
            forge.Load();
            Text = "🍳 DishForge - Генератор блюд";
            Size = new System.Drawing.Size(700, 550);
            StartPosition = FormStartPosition.CenterScreen;

            var top = new FlowLayoutPanel { Dock = DockStyle.Top, Padding = new Padding(5) };
            top.Controls.Add(new Label { Text = "Тип:", AutoSize = true });
            typeBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, Items = { "", "breakfast", "lunch", "dinner", "dessert", "snack", "drink" }, Width = 100 };
            top.Controls.Add(typeBox);
            top.Controls.Add(new Label { Text = "Кухня:", AutoSize = true });
            cuisineBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, Items = { "", "italian", "japanese", "russian", "mexican", "french", "chinese", "indian" }, Width = 100 };
            top.Controls.Add(cuisineBox);
            top.Controls.Add(new Label { Text = "Время (мин):", AutoSize = true });
            timeBox = new TextBox { Width = 60 };
            top.Controls.Add(timeBox);
            top.Controls.Add(new Label { Text = "Сложность:", AutoSize = true });
            diffBox = new ComboBox { DropDownStyle = ComboBoxStyle.DropDownList, Items = { "", "easy", "medium", "hard" }, Width = 80 };
            top.Controls.Add(diffBox);
            var genBtn = new Button { Text = "🎲 Сгенерировать" };
            genBtn.Click += (s, e) => Generate();
            top.Controls.Add(genBtn);
            var menuBtn = new Button { Text = "🍽️ Меню (3)" };
            menuBtn.Click += (s, e) => GenerateMenu(3);
            top.Controls.Add(menuBtn);
            Controls.Add(top);

            resultBox = new TextBox { Multiline = true, ReadOnly = true, ScrollBars = ScrollBars.Vertical, Dock = DockStyle.Fill };
            Controls.Add(resultBox);

            var bottom = new FlowLayoutPanel { Dock = DockStyle.Bottom, Padding = new Padding(5) };
            var favBtn = new Button { Text = "⭐ Добавить в избранное" };
            favBtn.Click += (s, e) => AddFavorite();
            bottom.Controls.Add(favBtn);
            var listFavBtn = new Button { Text = "📋 Избранное" };
            listFavBtn.Click += (s, e) => ShowFavorites();
            bottom.Controls.Add(listFavBtn);
            Controls.Add(bottom);
        }

        void Generate()
        {
            string type = typeBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(type)) type = null;
            string cuisine = cuisineBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(cuisine)) cuisine = null;
            string timeStr = timeBox.Text.Trim();
            int? maxTime = string.IsNullOrEmpty(timeStr) ? (int?)null : int.Parse(timeStr);
            string diff = diffBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(diff)) diff = null;
            var dish = forge.Generate(type, cuisine, maxTime, diff);
            if (dish != null)
            {
                currentDish = dish;
                resultBox.Text = forge.FormatDish(dish);
            }
            else
            {
                resultBox.Text = "❌ Блюда не найдены";
                currentDish = null;
            }
        }

        void GenerateMenu(int count)
        {
            string type = typeBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(type)) type = null;
            string cuisine = cuisineBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(cuisine)) cuisine = null;
            string timeStr = timeBox.Text.Trim();
            int? maxTime = string.IsNullOrEmpty(timeStr) ? (int?)null : int.Parse(timeStr);
            string diff = diffBox.SelectedItem.ToString();
            if (string.IsNullOrEmpty(diff)) diff = null;
            var menu = forge.GenerateMenu(count, type, cuisine, maxTime, diff);
            resultBox.Text = "";
            if (menu.Any())
            {
                resultBox.AppendText($"🍽️ Меню из {menu.Count} блюд:\n\n");
                for (int i = 0; i < menu.Count; i++)
                {
                    var d = menu[i];
                    resultBox.AppendText($"{i+1}. {d.Name} ({d.Type}, {d.Cuisine}, {d.Time} мин, {d.Difficulty})\n");
                    resultBox.AppendText($"   {d.Description}\n\n");
                }
            }
            else
            {
                resultBox.AppendText("❌ Недостаточно блюд");
            }
            currentDish = null;
        }

        void AddFavorite()
        {
            if (currentDish == null) { MessageBox.Show("Сначала сгенерируйте блюдо"); return; }
            if (forge.AddFavorite(currentDish.Name)) MessageBox.Show("Добавлено в избранное");
            else MessageBox.Show("Уже в избранном");
        }

        void ShowFavorites()
        {
            var favs = forge.GetFavorites();
            if (!favs.Any()) { MessageBox.Show("Избранное пусто"); return; }
            var dialog = new Form { Text = "⭐ Избранное", Size = new System.Drawing.Size(500, 400), StartPosition = FormStartPosition.CenterParent };
            var area = new TextBox { Multiline = true, ReadOnly = true, ScrollBars = ScrollBars.Vertical, Dock = DockStyle.Fill };
            favs.ForEach(d => area.AppendText($"🍽️ {d.Name}\n  Тип: {d.Type}, Кухня: {d.Cuisine}, Время: {d.Time} мин, Сложность: {d.Difficulty}\n\n"));
            dialog.Controls.Add(area);
            dialog.ShowDialog();
        }
    }
}

File: dish_forge.js
#!/usr/bin/env node
/**
 * dish_forge.js - Генератор случайных блюд на JavaScript (Node.js CLI + веб)
 */
const fs = require('fs');
const path = require('path');
const { program } = require('commander');

const DATA_FILE = path.join(__dirname, 'dishes.json');
const FAVORITES_FILE = path.join(__dirname, 'favorites.json');

class Dish {
    constructor(name, type, cuisine, time, difficulty, description, ingredients, instructions) {
        this.name = name;
        this.type = type;
        this.cuisine = cuisine;
        this.time = time;
        this.difficulty = difficulty;
        this.description = description;
        this.ingredients = ingredients;
        this.instructions = instructions;
    }
}

class DishForge {
    constructor() {
        this.dishes = [];
        this.favorites = [];
        this.load();
        this.loadFavorites();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            try {
                const data = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
                this.dishes = data.map(d => new Dish(d.name, d.type, d.cuisine, d.time, d.difficulty, d.description, d.ingredients, d.instructions));
                return;
            } catch {}
        }
        this.dishes = this.defaultDishes();
        this.save();
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.dishes, null, 2));
    }

    loadFavorites() {
        if (fs.existsSync(FAVORITES_FILE)) {
            try {
                this.favorites = JSON.parse(fs.readFileSync(FAVORITES_FILE, 'utf8'));
            } catch {}
        }
    }

    saveFavorites() {
        fs.writeFileSync(FAVORITES_FILE, JSON.stringify(this.favorites, null, 2));
    }

    defaultDishes() {
        return [
            new Dish("Овсянка с ягодами", "breakfast", "russian", 10, "easy",
                "Полезный завтрак из овсяных хлопьев с ягодами и мёдом.",
                ["Овсяные хлопья 100г", "Молоко 200мл", "Ягоды 50г", "Мёд 1 ст.л."],
                ["Вскипятить молоко", "Добавить хлопья, варить 5 мин", "Добавить ягоды и мёд"]),
            new Dish("Паста Карбонара", "lunch", "italian", 25, "medium",
                "Классическая итальянская паста с беконом и пармезаном.",
                ["Паста 200г", "Бекон 100г", "Яйца 2 шт", "Пармезан 50г", "Сливки 100мл"],
                ["Отварить пасту", "Обжарить бекон", "Смешать яйца, сыр и сливки", "Соединить с пастой"]),
            new Dish("Суши Филадельфия", "dinner", "japanese", 40, "hard",
                "Роллы с лососем, сливочным сыром и авокадо.",
                ["Рис для суши 200г", "Нори 4 шт", "Лосось 150г", "Сливочный сыр 100г", "Авокадо 1 шт"],
                ["Сварить рис", "Нарезать лосось и авокадо", "Собрать роллы", "Нарезать и подавать"]),
            new Dish("Борщ", "dinner", "russian", 60, "medium",
                "Традиционный украинский суп со свеклой и капустой.",
                ["Свекла 2 шт", "Капуста 200г", "Картофель 3 шт", "Мясо 300г", "Лук 1 шт", "Морковь 1 шт"],
                ["Сварить бульон", "Обжарить лук и морковь", "Добавить свеклу", "Соединить с бульоном и варить"]),
            new Dish("Чизкейк", "dessert", "french", 50, "medium",
                "Нежный творожный десерт с печеньем и ягодами.",
                ["Печенье 200г", "Масло 100г", "Творог 500г", "Сахар 150г", "Яйца 3 шт", "Сливки 200мл"],
                ["Измельчить печенье", "Смешать с маслом", "Взбить творог с сахаром и яйцами", "Запекать 40 мин"]),
        ];
    }

    generate(type, cuisine, maxTime, difficulty) {
        let filtered = this.dishes;
        if (type) filtered = filtered.filter(d => d.type === type);
        if (cuisine) filtered = filtered.filter(d => d.cuisine === cuisine);
        if (maxTime) filtered = filtered.filter(d => d.time <= maxTime);
        if (difficulty) filtered = filtered.filter(d => d.difficulty === difficulty);
        if (!filtered.length) return null;
        return filtered[Math.floor(Math.random() * filtered.length)];
    }

    generateMenu(count, filters = {}) {
        let available = this.dishes;
        if (filters.type) available = available.filter(d => d.type === filters.type);
        if (filters.cuisine) available = available.filter(d => d.cuisine === filters.cuisine);
        if (filters.maxTime) available = available.filter(d => d.time <= filters.maxTime);
        if (filters.difficulty) available = available.filter(d => d.difficulty === filters.difficulty);
        if (available.length < count) count = available.length;
        const shuffled = [...available].sort(() => Math.random() - 0.5);
        return shuffled.slice(0, count);
    }

    getTypes() { return [...new Set(this.dishes.map(d => d.type))].sort(); }
    getCuisines() { return [...new Set(this.dishes.map(d => d.cuisine))].sort(); }
    getDifficulties() { return [...new Set(this.dishes.map(d => d.difficulty))].sort(); }

    addFavorite(name) {
        if (!this.favorites.includes(name)) {
            this.favorites.push(name);
            this.saveFavorites();
            return true;
        }
        return false;
    }

    removeFavorite(name) {
        const idx = this.favorites.indexOf(name);
        if (idx !== -1) {
            this.favorites.splice(idx, 1);
            this.saveFavorites();
            return true;
        }
        return false;
    }

    getFavorites() {
        return this.dishes.filter(d => this.favorites.includes(d.name));
    }

    formatDish(dish) {
        let lines = [];
        lines.push(`🍽️ ${dish.name}`);
        lines.push(`  Тип: ${dish.type}`);
        lines.push(`  Кухня: ${dish.cuisine}`);
        lines.push(`  Время: ${dish.time} мин`);
        lines.push(`  Сложность: ${dish.difficulty}`);
        lines.push(`  Описание: ${dish.description}`);
        lines.push("  Ингредиенты:");
        dish.ingredients.forEach(ing => lines.push(`    - ${ing}`));
        lines.push("  Приготовление:");
        dish.instructions.forEach((step, i) => lines.push(`    ${i+1}. ${step}`));
        return lines.join("\n");
    }
}

program
    .command('generate')
    .option('--type <type>', 'Тип блюда')
    .option('--cuisine <cuisine>', 'Кухня')
    .option('--time <time>', 'Макс. время (мин)', parseInt)
    .option('--difficulty <difficulty>', 'Сложность')
    .action((options) => {
        const forge = new DishForge();
        const dish = forge.generate(options.type, options.cuisine, options.time, options.difficulty);
        if (dish) console.log(forge.formatDish(dish));
        else console.log('❌ Блюда не найдены');
    });

program
    .command('random-menu')
    .option('--count <count>', 'Количество блюд', parseInt, 3)
    .option('--type <type>', 'Тип блюда')
    .option('--cuisine <cuisine>', 'Кухня')
    .option('--time <time>', 'Макс. время', parseInt)
    .option('--difficulty <difficulty>', 'Сложность')
    .action((options) => {
        const forge = new DishForge();
        const filters = {};
        if (options.type) filters.type = options.type;
        if (options.cuisine) filters.cuisine = options.cuisine;
        if (options.time) filters.maxTime = options.time;
        if (options.difficulty) filters.difficulty = options.difficulty;
        const dishes = forge.generateMenu(options.count, filters);
        if (dishes.length) {
            console.log(`🍽️ Меню из ${dishes.length} блюд:`);
            dishes.forEach((d, i) => {
                console.log(`\n${i+1}. ${d.name} (${d.type}, ${d.cuisine}, ${d.time} мин, ${d.difficulty})`);
                console.log(`   ${d.description}`);
            });
        } else {
            console.log('❌ Недостаточно блюд');
        }
    });

program
    .command('list')
    .option('--type <type>', 'Фильтр по типу')
    .option('--cuisine <cuisine>', 'Фильтр по кухне')
    .action((options) => {
        const forge = new DishForge();
        let dishes = forge.dishes;
        if (options.type) dishes = dishes.filter(d => d.type === options.type);
        if (options.cuisine) dishes = dishes.filter(d => d.cuisine === options.cuisine);
        dishes.forEach(d => console.log(`${d.name} (${d.type}, ${d.cuisine}, ${d.time} мин)`));
    });

program
    .command('favorites')
    .command('add <name>').action((name) => {
        const forge = new DishForge();
        if (forge.addFavorite(name)) console.log(`✅ '${name}' добавлен в избранное`);
        else console.log(`⚠️ '${name}' уже в избранном или не существует`);
    })
    .command('remove <name>').action((name) => {
        const forge = new DishForge();
        if (forge.removeFavorite(name)) console.log(`✅ '${name}' удалён из избранного`);
        else console.log(`❌ '${name}' не найден в избранном`);
    })
    .command('list').action(() => {
        const forge = new DishForge();
        const favs = forge.getFavorites();
        if (favs.length) {
            console.log('⭐ Избранное:');
            favs.forEach(d => console.log(`  ${d.name} (${d.type}, ${d.cuisine})`));
        } else {
            console.log('Избранное пусто');
        }
    });

if (process.argv.length <= 2) {
    // interactive mode
    const forge = new DishForge();
    const readline = require('readline');
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const prompt = (q) => new Promise(resolve => rl.question(q, resolve));

    (async () => {
        while (true) {
            console.log('\n🍳 DishForge - Генератор блюд (интерактивный)');
            console.log('1. Сгенерировать случайное блюдо');
            console.log('2. Сгенерировать меню');
            console.log('3. Список всех блюд');
            console.log('4. Избранное');
            console.log('0. Выход');
            const choice = await prompt('Выберите действие: ');
            switch (choice.trim()) {
                case '0': rl.close(); return;
                case '1': {
                    const t = await prompt('Тип (Enter пропустить): ') || undefined;
                    const c = await prompt('Кухня (Enter пропустить): ') || undefined;
                    const timeStr = await prompt('Макс. время (мин, Enter пропустить): ');
                    const maxTime = timeStr ? parseInt(timeStr) : undefined;
                    const diff = await prompt('Сложность (easy/medium/hard, Enter пропустить): ') || undefined;
                    const dish = forge.generate(t, c, maxTime, diff);
                    if (dish) console.log(forge.formatDish(dish));
                    else console.log('❌ Блюда не найдены');
                    break;
                }
                case '2': {
                    const count = parseInt(await prompt('Количество блюд (по умолчанию 3): ') || '3');
                    const t = await prompt('Тип (Enter пропустить): ') || undefined;
                    const c = await prompt('Кухня (Enter пропустить): ') || undefined;
                    const timeStr = await prompt('Макс. время (мин, Enter пропустить): ');
                    const maxTime = timeStr ? parseInt(timeStr) : undefined;
                    const diff = await prompt('Сложность (Enter пропустить): ') || undefined;
                    const filters = {};
                    if (t) filters.type = t;
                    if (c) filters.cuisine = c;
                    if (maxTime) filters.maxTime = maxTime;
                    if (diff) filters.difficulty = diff;
                    const dishes = forge.generateMenu(count, filters);
                    if (dishes.length) {
                        console.log(`🍽️ Меню из ${dishes.length} блюд:`);
                        dishes.forEach((d, i) => {
                            console.log(`\n${i+1}. ${d.name} (${d.type}, ${d.cuisine}, ${d.time} мин, ${d.difficulty})`);
                            console.log(`   ${d.description}`);
                        });
                    } else {
                        console.log('❌ Недостаточно блюд');
                    }
                    break;
                }
                case '3': {
                    const t = await prompt('Тип (Enter пропустить): ') || undefined;
                    const c = await prompt('Кухня (Enter пропустить): ') || undefined;
                    let dishes = forge.dishes;
                    if (t) dishes = dishes.filter(d => d.type === t);
                    if (c) dishes = dishes.filter(d => d.cuisine === c);
                    dishes.forEach(d => console.log(`  ${d.name} (${d.type}, ${d.cuisine}, ${d.time} мин)`));
                    break;
                }
                case '4': {
                    const favs = forge.getFavorites();
                    console.log('⭐ Избранное:');
                    if (favs.length) favs.forEach(d => console.log(`  ${d.name} (${d.type}, ${d.cuisine})`));
                    else console.log('  Пусто');
                    console.log('\n1. Добавить в избранное');
                    console.log('2. Удалить из избранного');
                    console.log('3. Назад');
                    const sub = await prompt('Выберите: ');
                    if (sub === '1') {
                        const name = await prompt('Название блюда: ');
                        if (forge.addFavorite(name)) console.log(`✅ '${name}' добавлен`);
                        else console.log(`⚠️ '${name}' уже в избранном или не существует`);
                    } else if (sub === '2') {
                        const name = await prompt('Название блюда: ');
                        if (forge.removeFavorite(name)) console.log(`✅ '${name}' удалён`);
                        else console.log(`❌ '${name}' не найден`);
                    }
                    break;
                }
                default: console.log('Неверный выбор');
            }
        }
    })();
} else {
    program.parse(process.argv);
}
